### ISSUE TITLE
Something is wrong with array signatures

### ISSUE THREAD

**Issue by forki** [ionide/ionide-vscode-fsharp#1240]
![image](https://user-images.githubusercontent.com/57396/66600897-e32d3500-eba6-11e9-805d-c3122e75b4d9.png)

**Comment by Krzysztof-Cieslak** [comment]
CC: @7sharp9 ;-)

**Comment by 7sharp9** [comment]
Easy solution, just don’t use arrays 👾

**Comment by baronfel** [comment]
Also any type that needs to be qualified (either because its parent module isn't open in the current scope or because the module is RequireQualifiedAccess):

![type of name `Structure.Scope` is rendered as '``Structure.Scope``'](https://user-images.githubusercontent.com/573979/66608877-2bcff880-eb7d-11e9-906f-b9160253e70c.png)

**Comment by 7sharp9** [comment]
I wonder is require qualified access is a property of the symbol

**Comment by forki** [comment]
![image](https://user-images.githubusercontent.com/57396/66909128-84cfde80-f00c-11e9-9261-5da0aeddcf3f.png)

especially annoying when inconsistent

**Comment by forki** [comment]
The compile is right.

**Comment by 7sharp9** [comment]
What I mean is the compiler represents generics in 3 different way,
consistency = None

**Comment by forki** [comment]
The problem is not the compiler treating some generics differently. The problem is that tooling somehow regressed. Does anyone already know why?

**Comment by 7sharp9** [comment]
Yep, FsharpType.Format has a issues where it deals with double ticks
indistinctly, where possible uses of it needs to be replaced like the other
tooltip construction parts.

**Comment by Krzysztof-Cieslak** [comment]
https://github.com/fsharp/FsAutoComplete/commit/cd2b1428ae721040f9c47e97b8df3825f9e1cec5#diff-4f277f322ee142a567b25f775b749079R36 - this seems to always force `typ<args>` format

**Comment by 7sharp9** [comment]
I thought this was the culprit too:
https://github.com/fsharp/FsAutoComplete/commit/cd2b1428ae721040f9c47e97b8df3825f9e1cec5#diff-4f277f322ee142a567b25f775b749079R41
As the format may be a longident and always surrounded with double ticks

**Comment by Krzysztof-Cieslak** [comment]
> I thought this was the culprit too:

Yes, this is probably causing the case which @baronfel reported - we probably can do some heuristics here - if `.` is only symbol we use in name we shouldn't pretty-print it - something like https://github.com/fsharp/FsAutoComplete/blob/master/src/FsAutoComplete/CommandResponse.fs#L577

For the option/array/list... I guess we can just do bunch of if/else to handle those few types that are formatted differently by compiler /shrug

**Comment by forki** [comment]
The FCS does not give us a DisplayEnv to properly format?

**Comment by 7sharp9** [comment]
FSharpType.Format gives no control to the formatting so where possible the
custom construction of tips or docs should be used just as it is in the
more fancy tooltips.

**Comment by 7sharp9** [comment]
First what would these ideally show in a tip:
```
let arrayThing1 = [||]
let arrayThing2 : array string= [||]
let arrayThing3 : string [] = [||]
```
FCS will format these all differently:
```
'a []
string array
string []
```

**Comment by 7sharp9** [comment]
```
let arraything : array<'a> = [||]
```
is 
```
'a array
```

**Comment by Krzysztof-Cieslak** [comment]
>  I think the intent of the code in the tips is to always format consistently as

I think this was your change, in the *tooltips* previously we've used Formats representation.

We do normalize to `generic<types>` in *Info Panel* to get better linking support

I think in *tooltips* we should use `typ array` for array,list, lazy, async, option, seq, set (any more?)

**Comment by 7sharp9** [comment]
With arrays they have special support in the type definition `IsArrayType` so what I'll do is deal with those first and we can build this up in stages.  One of the problems is Format does not correctly format double tick idents properly.  All the info is present to format them which is what lead to the custom tooltips so its just a matter of dealing with each case.

**Comment by 7sharp9** [comment]
Currently the signatures with generics all end up the same %s<%types>

**Comment by 7sharp9** [comment]
Because we potentially we have the TypeDefinition which is an `FsharpEntity` we could use `getEntitySignature` to format, sometimes those there is no TypeDefinition though so `HasTypeDefinition` needs to be checked first.

**Comment by 7sharp9** [comment]
#482 is now WIP

**Comment by 7sharp9** [comment]
Fixes are in 482 now.

**Comment by 7sharp9** [comment]
https://github.com/fsharp/FsAutoComplete/pull/482
wrong repo :-p

**Comment by charlieturndorf** [comment]
At the same time I started seeing ``` ``[]``<'T>``` instead of `` 'T [] ``, I also started seeing ``` float<MeasureProduct<MeasureInverse<u>,MeasureProduct<q,``1``>>> ``` instead of ``` float<q/u> ```. This is quite hard to read, and I'm working in a measure-heavy codebase so I look at a lot of measures.

**Comment by 7sharp9** [comment]
I think this is fixed in the link above, otherwise its just another PR away :-)

**Comment by Krzysztof-Cieslak** [comment]
This should be fixed in `4.3.0`
